addon_service/credentials/encryption.py

```diff
@@ -0,0 +1,142 @@
+"""for encrypting and decrypting data (using secrets from settings)
+
+considering recommendations from:
+- https://cryptography.io/en/latest/fernet/
+- https://docs.python.org/3.12/library/hashlib.html#hashlib.scrypt
+- https://cryptography.io/en/latest/hazmat/primitives/key-derivation-functions/#scrypt
+- https://datatracker.ietf.org/doc/html/rfc7914
+- https://nvlpubs.nist.gov/nistpubs/Legacy/SP/nistspecialpublication800-132.pdf
+"""
+
+import base64
+import dataclasses
+import functools
+import hashlib
+import json
+import math
+import os
+
+from cryptography import fernet
+from django.conf import settings
+
+
+__all__ = (
+    "pls_decrypt_bytes",
+    "pls_decrypt_json",
+    "pls_encrypt_bytes",
+    "pls_encrypt_json",
+    "salt_factory",
+)
+
+
+# 32-byte key expected by cryptography.fernet.Fernet
+_KEY_BYTE_COUNT = 32
+
+# recommended len(salt) >= 16 bytes
+_SALT_BYTE_COUNT = settings.GRAVYVALET_SALT_BYTE_COUNT or 17
+
+
+def salt_factory() -> bytes:
+    return os.urandom(_SALT_BYTE_COUNT)
+
+
+@dataclasses.dataclass(frozen=True)  # frozen for use as lru_cache key
+class KeyParameters:  # https://datatracker.ietf.org/doc/html/rfc7914#section-2
+    salt: bytes = dataclasses.field(default_factory=salt_factory)
+    # recommended scrypt_cost ("N") between 2^14 and 2^20
+    scrypt_cost: int = settings.GRAVYVALET_SCRYPT_COST or 2**17
+    # recommended scrypt_block_size ("r") = 8
+    scrypt_block_size: int = settings.GRAVYVALET_SCRYPT_BLOCK_SIZE or 8
+    # recommended scrypt_parallelization ("p") = 1
+    scrypt_parallelization: int = settings.GRAVYVALET_SCRYPT_PARALLELIZATION or 3
+
+    def __post_init__(self):
+        assert self.scrypt_block_size > 1
+        # scrypt_cost "must be larger than 1, a power of 2, and less than 2^(128 * r / 8)"
+        _cost_log2 = math.log2(self.scrypt_cost)
+        assert (
+            self.scrypt_cost > 1
+            and _cost_log2 == int(_cost_log2)
+            and _cost_log2 <= (128 * self.scrypt_block_size / 8)
+        )
+        # scrypt_parallelization "is a positive integer less than or equal to ((2^32-1) * 32) / (128 * r)"
+        assert self.scrypt_parallelization <= ((2**32 - 1) * 32) / (
+            128 * self.scrypt_block_size
+        )
+
+    def memory_required(self) -> int:
+        return self.scrypt_cost * self.scrypt_block_size * 129
```

Comment:
Why `* 129`? I am sure there is a reason. I am simply curious.

Reply:
i read somewhere that scrypt's approximate memory use is `N * r * 128` bytes, and fudged it up to 129 for `hashlib.scrypt` to stop throwing "memory limit exceeded" errors

...but now i've verified that estimate (it's a lower bound for any scrypt implementation not trying to play games recomputing intermediate values) and updated this function with explanatory comments and more empirically tested fudge